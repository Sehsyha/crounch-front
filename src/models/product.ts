export interface ProductResponse {
  products: ProductApi[];
}

export interface ProductApi {
  product_name: string;
  categories: string;
  code: string;
}

export interface Product {
  barCode: string;
  name: string;
  categories: string[];
}
