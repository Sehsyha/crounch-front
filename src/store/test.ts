import { ActionTree, MutationTree, GetterTree, Module, ActionContext } from 'vuex';
import { RootState } from '.';

export type ActionCall<S, R> = (injectee: ActionContext<S, R>, payload?: any) => any;

export function createStoreModuleMock<S, A, M, G>(initialState: S, actionKeys: A, mutationKeys: M, getterKeys: G): Module<S, RootState> {
  const actions: ActionTree<S, RootState> = {};
  const mutations: MutationTree<S> = {};
  const getters: GetterTree<S, RootState> = {};

  Object.keys(actionKeys).forEach(key => actions[key] = jest.fn());
  Object.keys(mutationKeys).forEach(key => mutations[key] = jest.fn());
  Object.keys(getterKeys).forEach(key => getters[key] = jest.fn());

  return {
    namespaced: true,
    actions,
    getters,
    mutations,
    state: initialState,
  }
}
