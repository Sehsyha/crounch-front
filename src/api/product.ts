import { doFetch } from './api';
import { Product } from '@/models/product';

export function createProduct(name: string): Promise<Product> {
  return doFetch({
    url: 'products',
    method: 'POST',
    data: { name }
  });
}

export function searchProduct(name: string): Promise<Array<Product>> {
  return doFetch({
    url: 'products/search',
    method: 'POST',
    data: { name }
  })
}
