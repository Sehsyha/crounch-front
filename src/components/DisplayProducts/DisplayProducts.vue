<template>
  <div>
    <h3>Produits</h3>
    <ul v-if="products.length" id="list-products">
      <li v-for="product in products" :key="product.id">{{ product.name }}</li>
    </ul>
    <p v-else>Aucun produits dans cette liste</p>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Prop } from 'vue-property-decorator';
import { Product } from '../../models/product';

@Component
export default class DisplayProducts extends Vue {
  @Prop()
  public products!: Product[]
}
</script>
