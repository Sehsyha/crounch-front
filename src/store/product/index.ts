import { Module } from 'vuex';

import { actions } from './actions';
import { getters } from './getters';
import { mutations } from './mutations';
import { RootState } from '..';
import { Product } from '@/models/product';

export interface ProductState {
  products: Product[];
}

export const state: ProductState = {
  products: [],
};

const namespaced: boolean = true;
export const productNamespace = { namespace: 'product' };

export const product: Module<ProductState, RootState> = {
  namespaced,
  state,
  actions,
  getters,
  mutations,
};
