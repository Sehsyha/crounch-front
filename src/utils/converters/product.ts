import { ProductApi, Product } from '@/models/product';

export function convertProductFromApi(productsApi: ProductApi[]): Product[] {
  return productsApi.map(productApi => ({
    name: productApi.product_name,
    categories: productApi.categories.replace(' ', '').split(','),
    barCode: productApi.code,
  }));
}
