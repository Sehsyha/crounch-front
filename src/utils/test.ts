import Vue from 'vue';
import VueRouter from 'vue-router';

import Vuex, { ActionContext, ActionTree, ModuleTree } from 'vuex';
import { shallowMount, VueClass, createLocalVue, Slots } from '@vue/test-utils';

import { RootState, ActionCall } from '@/store';

export interface ShallowOptions {
  values?: object;
  modules?: ModuleTree<RootState>;
  slots?: Slots;
}

const emptyOptions: ShallowOptions = {
  values: {},
  modules: {},
  slots: {}
};

export function shallowComponent<V extends Vue>(component: VueClass<V>, { values, modules, slots }: ShallowOptions = emptyOptions) {
  const localVue = createLocalVue();

  localVue.use(Vuex);

  const router = new VueRouter();

  const store = new Vuex.Store<RootState>({
    modules,
  });

  return shallowMount(component, {
    attachToDocument: true,
    propsData: {
      ...values,
    },
    slots,
    store,
    router,
    localVue,
    stubs: ['router-link', 'router-view'],
    mocks: {
      $router: {
        push: jest.fn()
      }
    }
  });
}

export const callAction = <S>(context: ActionContext<S, RootState>, actions: ActionTree<S, RootState>) => (
  key: string,
  params?: any
): any => {
  return (actions[key] as ActionCall<S, RootState>)(context, params);
};

export function initContext<S>(context: Partial<ActionContext<S, RootState>>): ActionContext<S, RootState> {
  const emptyContext = createEmptyActionContext<S>();

  return {
    ...emptyContext,
    ...context,
  };
}

function createEmptyActionContext<S>(): ActionContext<S, RootState> {
  return {
    state: {},
    getters: {},
    dispatch: {},
    commit: {},
    rootState: {
      dummy: '',
    },
    rootGetters: {},
  } as ActionContext<S, RootState>;
}
