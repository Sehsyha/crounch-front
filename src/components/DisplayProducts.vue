<template>
  <ul v-if="products.length" id="list-products">
    <DisplayProduct
      v-for="product in orderedProducts"
      :key="product.id"
      :product="product"
      :is-selected-list-archive="isSelectedListArchived"
      @delete-product="deleteProduct"
      @toggle-bought-product="toggleBoughtProduct"
    />
  </ul>
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Emit, Prop } from 'vue-property-decorator'
import { getModule } from 'vuex-module-decorators'

import { Events } from '@/utils/events'
import { ProductInSelectedList } from '@/models/product'
import { ListModule } from '@/store/list/ListModule'
import { SetBoughtProductActionPayload } from '@/store/list/payloads'
import DisplayProduct from '@/ui/DisplayProduct.vue'

@Component({
  components: {
    DisplayProduct
  }
})
export default class DisplayProducts extends Vue {
  public listModule: ListModule = getModule(ListModule)

  @Prop()
  public products!: ProductInSelectedList[];

  get isSelectedListArchived() {
    return this.listModule.isSelectedListArchived
  }

  get orderedProducts () {
    const orderedProducts: ProductInSelectedList[] = []
    this.products.forEach(product => {
      if (product.bought) {
        orderedProducts.push(product)
      } else {
        orderedProducts.unshift(product)
      }
    })

    return orderedProducts
  }

  deleteProduct(product: ProductInSelectedList) {
    this.listModule.deleteProductAction(product)
  }

  toggleBoughtProduct(product: ProductInSelectedList) {
    const setBoughtProductActionPayload: SetBoughtProductActionPayload = {
      product,
      bought: !product.bought,
    }
    this.listModule.setBoughtProductAction(setBoughtProductActionPayload)
  }
}
</script>
