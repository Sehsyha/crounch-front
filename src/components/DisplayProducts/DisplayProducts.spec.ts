import { shallowComponent } from '@/utils/test'
import DisplayProducts from './DisplayProducts.vue'
import { Product } from '@/models/product'

describe('DisplayProducts', () => {
  it('Should render a message when there is no products.', () => {
    const values = {
      products: [],
    }

    const wrapper = shallowComponent(DisplayProducts, { values })

    expect(wrapper).toMatchSnapshot()
  })

  it('Should render the products list when there is products.', () => {
    const products: Product[] = [
      {
        id: 'product id',
        name: 'product name'
      }
    ]
    const values = {
      products
    }

    const wrapper = shallowComponent(DisplayProducts, { values })

    expect(wrapper).toMatchSnapshot()
  })
})
