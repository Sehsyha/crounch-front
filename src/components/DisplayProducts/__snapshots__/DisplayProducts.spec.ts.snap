// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayProducts Should render a message when there is no products. 1`] = `
<div>
  <h3>Produits</h3>
  <p>Aucun produits dans cette liste</p>
</div>
`;

exports[`DisplayProducts Should render the products list when there is products. 1`] = `
<div>
  <h3>Produits</h3>
  <ul>
    <li>product name</li>
  </ul>
</div>
`;
